Added FileIO Support for ORC Reader and Writers

pavibhai opened the pull request:
## What?
This adds support for FileIO with Apache ORC by wrapping InputFile and OutputFile into a consumable FileSystem objects.

## Why?
In the absence of this ORC is not able to leverage the benefits of better FileIO such as S3FileIO compared to S3AFileSystem

## Tested?
New unit tests have been added to ensure the use of FileIO in ORC Readers and Writers

nastra commented on orc/src/test/java/org/apache/iceberg/orc/TestOrcDataWriter.java:
```diff
@@ -126,4 +135,125 @@ public void testDataWriter() throws IOException {
+    // Show that FileSystem access is not possible for the file we are supplying as the scheme
+    // dummy is not handled
+    ProxyOutputFile outFile = new ProxyOutputFile(Files.localOutput(temp.newFile()));
+    Assertions.assertThrows(
```
it would be better to move such assertions to AssertJ, because otherwise it'll be more work potentially to move such code when upgrading to Junit5

pavibhai replied:
Makes sense. Changing.

RussellSpitzer commented:
I'm not a big fan of the fake filesystem approach here, mostly because i'm afraid of mocking an object like that when we don't have the full filesystem state. I feel like this patch would have us maintaining a rather large Hadoop mock.

Is there any chance we can convince the ORC project to allow the creation of a writer from an "java.io.OutputStream" instead of always creating its own file?

rdblue commented on orc/src/main/java/org/apache/iceberg/orc/ORC.java:
```diff
@@ -789,7 +808,210 @@ static Reader newFileReader(InputFile file, Configuration config) {
+
+    @Override
+    public FSDataOutputStream create(Path f) throws IOException {
+      return create(f, null, true, 0, (short) 0, 0, null);
```
Like above, I this is the method to implement, not `create(...)`.

pavibhai replied:
Agreed

rdblue commented on orc/src/main/java/org/apache/iceberg/orc/ORC.java:
```diff
@@ -789,7 +808,210 @@ static Reader newFileReader(InputFile file, Configuration config) {
+    }
+  }
+
+  private static class NullFileSystem extends FileSystem {
```
Is there a better place for this fake FS stuff? Maybe a top-level util class?

pavibhai replied:
Sure. Moved to FileIOFSUtil class

rdblue commented on orc/src/main/java/org/apache/iceberg/orc/ORC.java:
```diff
@@ -789,7 +808,210 @@ static Reader newFileReader(InputFile file, Configuration config) {
+    @Override
+    protected void finalize() throws Throwable {
+      super.finalize();
+      if (!closed) {
```
I don't think there's a need for this since the input stream should have its own finalizer.

pavibhai replied:
Makes sense. Removing this.

rdblue commented on orc/src/main/java/org/apache/iceberg/orc/ORC.java:
```diff
@@ -789,7 +808,210 @@ static Reader newFileReader(InputFile file, Configuration config) {
     if (file instanceof HadoopInputFile) {
       readerOptions.filesystem(((HadoopInputFile) file).getFileSystem());
+    } else {
+      readerOptions.filesystem(new InputFileSystem(file)).maxLength(file.getLength());
```
`maxLength` is used to avoid calls to get status? If so, we should add a comment.

pavibhai replied:
Added comment to clarify this.

rdblue commented on orc/src/main/java/org/apache/iceberg/orc/ORC.java:
```diff
@@ -789,7 +808,210 @@ static Reader newFileReader(InputFile file, Configuration config) {
     ReaderOptions readerOptions = OrcFile.readerOptions(config).useUTCTimestamp(true);
     if (file instanceof HadoopInputFile) {
       readerOptions.filesystem(((HadoopInputFile) file).getFileSystem());
+    } else {
```
The `newFileReader(String, ReaderOptions)` above should be deprecated, right?

pavibhai replied:
I moved the OrcFile.createReader method into the new method and deprecated the previous one. That is a package local method, if acceptable I can remove it as it has no other uses.

rdblue replied:
Yeah, rather than deprecating we should just remove it if it isn't public.

pavibhai replied:
deleted

rdblue commented on orc/src/main/java/org/apache/iceberg/orc/ORC.java:
```diff
@@ -789,7 +808,210 @@ static Reader newFileReader(InputFile file, Configuration config) {
+    private boolean closed;
+    private final StackTraceElement[] createStack;
+
+    private WrappedSeekableInputStream(SeekableInputStream inputStream) {
```
Can you move this stream to HadoopStreams? There's nothing ORC specific here and we may find other uses for it. That would also allow us to update the `wrap` methods that convert to detect double wrapping and return the underlying stream.

pavibhai replied:
Did the following:
* Moved class to HadoopStreams
* Made HadoopStreams public to be accessible here
* Implemented DelegatingInputStream

rdblue commented on orc/src/main/java/org/apache/iceberg/orc/OrcFileAppender.java:
```diff
@@ -88,8 +86,7 @@
     options.setSchema(orcSchema);
-    this.writer = newOrcWriter(file, options, metadata);
-
+    this.writer = ORC.newFileWriter(file, options, metadata);
```
is the `options.fileSystem` call still needed above?

pavibhai replied:
Yes that is still needed. We continue to retain the current behavior for HadoopIO. So the FileSystem from HadoopOutputFile or HadoopInputFile is used for the FS operations. When we use anything other than HadoopIO the FileIOFS kicks in.

rdblue commented on orc/src/test/java/org/apache/iceberg/orc/TestOrcDataWriter.java:
```diff
@@ -126,4 +135,116 @@ public void testDataWriter() throws IOException {
+        .hasMessageStartingWith("Failed to get file system for path: dummy");
+
+    // We are creating the proxy
+    SortOrder sortOrder = SortOrder.builderFor(SCHEMA).withOrderId(10).asc("id").build();
```
I don't think the sort order is needed.

pavibhai replied:
Sure, removed it to simplify the test

rdblue commented on orc/src/test/java/org/apache/iceberg/orc/TestOrcDataWriter.java:
```diff
@@ -126,4 +135,116 @@ public void testDataWriter() throws IOException {
+            .schema(SCHEMA)
+            .createWriterFunc(GenericOrcWriter::buildWriter)
+            .overwrite()
+            .withSpec(PartitionSpec.unpartitioned())
```
This should be the default, right?

pavibhai replied:
Sorry didn't follow this. What do we mean by default?

Before the patch if you gave a Local[Output|Input]File this was converted to FileSystem operation resulting in a LocalFileSystem for handling this.
To ensure that is not taking place, we are mimicking a schema `dummy` that is not handled, so if any FS operations happen normally then it will fail with `Failed to get file system for path: dummy` but if handled via FileIOFS then it should be successful.

I will add this comment to the test to make it clearer.

rdblue replied:
Do we need to add `withSpec` or will it use the unpartitioned spec by default?

pavibhai replied:
yeah we need it, without it I get an error that `Cannot create data writer without spec`

rdblue commented on orc/src/test/java/org/apache/iceberg/orc/TestOrcDataWriter.java:
```diff
@@ -126,4 +135,116 @@ public void testDataWriter() throws IOException {
+    OrcFile.ReaderOptions options =
+        OrcFile.readerOptions(new Configuration())
+            .filesystem(new ORC.InputFileSystem(outFile.toInputFile()))
+            .maxLength(outFile.toInputFile().getLength());
```
I think this should use the `newFileReader` method rather than creating options here.

rdblue commented:
Looks mostly good overall. Thanks for getting this working @pavibhai!

pavibhai commented:
> I'm not a big fan of the fake filesystem approach here, mostly because i'm afraid of mocking an object like that when we don't have the full filesystem state. I feel like this patch would have us maintaining a rather large Hadoop mock.
> 
> Is there any chance we can convince the ORC project to allow the creation of a writer from an "java.io.OutputStream" instead of always creating its own file?

Thanks for this insight. Overall I agree that this should not be our long term answer and we shall work in the ORC community to make this integration better.
In the meantime hopefully this provides the capability and given the relative stability of the FileSystem APIs and limited exposure to just creation of the input and output streams hope this is acceptable.

pavibhai commented on orc/src/main/java/org/apache/iceberg/orc/ORC.java:
```diff
@@ -789,7 +808,210 @@ static Reader newFileReader(InputFile file, Configuration config) {
+
+    @Override
+    public FSDataInputStream open(Path f) throws IOException {
+      return open(f, 0);
```
Agreed this is sleek. Will change.

pavibhai commented on orc/src/main/java/org/apache/iceberg/orc/ORC.java:
```diff
@@ -789,7 +808,210 @@ static Reader newFileReader(InputFile file, Configuration config) {
+    }
+
+    @Override
+    public FSDataOutputStream create(Path f) throws IOException {
```
We do have that check, with the previous comment the check will now move to `create(Path f, boolean overwrite) method with the other method calling this.

pavibhai commented:
> Looks mostly good overall. Thanks for getting this working @pavibhai!

Thanks @rdblue for your comments. I have addressed the comments, there are a few comments where I gave additional clarification.

gaborkaszab commented on orc/src/main/java/org/apache/iceberg/orc/ORC.java:
```diff
@@ -85,10 +86,14 @@
 
 @SuppressWarnings("checkstyle:AbbreviationAsWordInName")
 public class ORC {
+  private static final Logger LOG = LoggerFactory.getLogger(ORC.class);
```
nit: I don't see any usage for this new Logger introduced by this patch.

rdblue commented on orc/src/main/java/org/apache/iceberg/orc/FileIOFSUtil.java:
```diff
@@ -0,0 +1,164 @@
+import org.apache.iceberg.io.OutputFile;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+
+public class FileIOFSUtil {
```
I don't think this should be public. Could you make it package-private?

pavibhai replied:
Sure, done.

rdblue commented:
Thanks, @pavibhai! Great to have this fixed, even if it's a hack :sweat_smile:.
